include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -377,45 +386,75 @@ __radix_sort_count_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments, :
         __hdl.depends_on(__dependency_event);
-
         // an accessor with value counter from each work_group
         oneapi::dpl::__ranges::__require_access(__hdl, __count_rng); //get an access to data under SYCL buffer
+        __hdl.parallel_for<_KernelName1>(
+#if _ONEDPL_COMPILE_KERNEL
+            __kernel_1,
+#endif
+            sycl::nd_range<1>(__radix_states * __scan_wg_size, __scan_wg_size), [=](sycl::nd_item<1> __self_item) {
+                // find borders of a region with a specific bucket id
+                sycl::global_ptr<_CountT> __begin = __count_rng.begin() + __scan_size * __self_item.get_group(0);
+                sycl::global_ptr<_CountT> __end = __begin + __scan_size;
 
-        __hdl.parallel_for<_KernelName>(
+                sycl::ONEAPI::exclusive_scan(__self_item.get_group(), __begin, __end, __begin, _CountT(0),
+                                             sycl::ONEAPI::plus<_CountT>{});
+            });
+    });
+
+    // 2. Global scan: produces global offsets using local offsets
+    __scan_event = __exec.queue().submit([&](sycl::handler& __hdl) {
+        __hdl.depends_on(__scan_event);
+        oneapi::dpl::__ranges::__require_access(__hdl, __count_rng);
+        __hdl.parallel_for<_KernelName2>(
+#if _ONEDPL_COMPILE_KERNEL
+            __kernel_2,
+#endif
             sycl::nd_range<1>(__radix_states, __radix_states), [=](sycl::nd_item<1> __self_item) {
-                // item info
-                const ::std::size_t __self_lidx = __self_item.get_local_id(0);
+                ::std::size_t __self_lidx = __self_item.get_local_id(0);
 
-                // exclusive scan (but in last __radix_states we write total sum)
-                ::std::uint32_t __part_sum = 0u;
-                for (::std::uint32_t __i = 0; __i <= __segments; ++__i)
-                {
-                    ::std::size_t __count_val_idx = __i * __radix_states + __self_lidx;
-                    ::std::uint32_t __prev_val = __count_rng[__count_val_idx];
-                    __count_rng[__count_val_idx] = __part_sum;
-                    __part_sum += __prev_val;
-                }
+                ::std::size_t __global_offset_idx = __global_scan_begin + __self_lidx;
+                ::std::size_t __last_segment_bucket_idx = (__self_lidx + 1) * __scan_size - 1;
 
-                // exclusive scan over total radix sums
-                ::std::size_t __total_sum_idx = __segments * __radix_states + __self_lidx;
-                __count_rng[__total_sum_idx] = sycl::ONEAPI::exclusive_scan(
-                    __self_item.get_group(), __count_rng[__total_sum_idx], sycl::ONEAPI::plus<_CountT>());
+                // copy buckets from the last segment, scan them to get global offsets
+                _CountT __val = __count_rng[__last_segment_bucket_idx];
+                __count_rng[__global_offset_idx] =
+                    sycl::ONEAPI::exclusive_scan(__self_item.get_group(), __val, sycl::ONEAPI::plus<_CountT>{});
```

Comment:
The size of the task is small, the time it takes should be negligible comparing to overhead coming from submitting the task. I do not see any way we can get rid of the task submit. On the other hand, using `sycl::ONEAPI::exclusive_scan` summaries what we do here, that results in better readability.

However, I will experiment with single_task to see if it can save some time.

Reply:
Changing the second scan kernel to the following:
```c++
        __hdl.single_task<_KernelName2>([=]() {
            _CountT __nex_val = 0;
            for(::std::size_t __radix_state = 0; __radix_state < __radix_states; ++__radix_state) {
                ::std::size_t __global_offset_idx = __global_scan_begin + __radix_state;
                ::std::size_t __last_segment_bucket_idx = (__radix_state + 1) * __scan_size - 1;

                // copy buckets from the last segment, scan them to get global offsets
                __count_rng[__global_offset_idx] = __nex_val;
                __nex_val += __count_rng[__last_segment_bucket_idx];
            }
        });
```
does not have significant impact on overall execution time on different devices.